### PR TITLE
Add AWS WAF Regional support for AWS ALBs

### PR DESCRIPTION
**Pull request by bje**
##### SUMMARY
Add AWS WAF Regional support for AWS ALBs.

The existing AWS WAF module only works with AWS Cloudfront and not with ALBs.  Instead of extending the existing AWS WAF module, I created new aws_waf_regional* components, since AWS themselves did not extend the `aws waf` CLI tool to support AWS WAF Regional, but instead added a `aws waf-regional` CLI option.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
- aws_waf_regional_condition
- aws_waf_regional_facts
- aws_waf_regional_rule
- aws_waf_regional_web_acl

##### ANSIBLE VERSION

```
ansible 2.6.2 (v2.6.2-alb-h2-waf-regional cfd25bf55b) last updated 2018/08/30 12:36:19 (GMT +100)
  config file = /Users/bje/devel/launchdarkly/playbooks/ansible.cfg
  configured module search path = [u'/Users/bje/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/bje/devel/launchdarkly/playbooks/src/ansible/lib/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.15 (default, Jun 17 2018, 12:46:58) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]
```

### CODE REVIEW

**Comment by mjmayer** [comment]
@bje I understand your justification for creating an entire new module for the regional WAF. After a cursory look at botocore difference between WAF and WAFRegional, the only method difference I saw was associate_web_acl. It seems like it would be easier to add that method to the existing Ansible WAF modules, rather than creating new modules and that are nearly identical to the existing WAF Module.

File: lib/ansible/modules/cloud/amazon/aws_waf_regional_rule.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    state:
+        description: whether the rule should be present or absent
+        choices:
+        - present
+        - absent
+        default: present
+    conditions:
+        description: >
+          list of conditions used in the rule. Each condition should
+          contain I(type): which is one of [C(byte), C(geo), C(ip), C(size), C(sql) or C(xss)]
```
Comment (gundalow) [inline lib/ansible/modules/cloud/amazon/aws_waf_regional_rule.py:44]:
```suggestion
          contain I(type) which is one of [C(byte), C(geo), C(ip), C(size), C(sql) or C(xss)]
```

File: lib/ansible/modules/cloud/amazon/aws_waf_regional_rule.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+          I(negated): whether the condition should be negated, and C(condition),
```
Comment (gundalow) [inline lib/ansible/modules/cloud/amazon/aws_waf_regional_rule.py:45]:
```suggestion
          I(negated) whether the condition should be negated, and C(condition),
```

The `: ` is what's breaking yaml parsing

**Comment by davehewy** [comment]
Is there going to be any more movement on this PR?
@mjmayer 's comments about preference perhaps for an optional flag on the existing waf_lib might be quicker?

**Comment by mjmayer** [comment]
@DaveHewy I'm working on a branch that uses the existing WAF modules. I will start a new PR once I'm done and link to it here. I'm guessing bje's PR isn't going to see any movement.

**Comment by davehewy** [comment]
@mjmayer as am i.
got an immediate need for it.

all i really needed to do was pass the resource=computed_resource to the boto client instantiation & add a module param to toggle waf-regional on and off. short-sighted i know, but for now less duplication. did you come up with similar solution? 

its also worth noting - i found no need to toy around with the region. since the module.param['region'] can just be used for both purposes

check that: i'm also seeing a difference in the `list-rules` method

**Comment by mjmayer** [comment]
@DaveHewy Yes, I came up with a similar solution. You will also need to add a  waiter for WAFRegional

https://github.com/mjmayer/ansible/blob/d665758be92b49209278b99904d8a72f3b026401/lib/ansible/module_utils/aws/waiters.py#L308-L313

**Comment by davehewy** [comment]
@mjmayer actual error
`An exception occurred during task execution. To see the full traceback, use -vvv. The error was: OperationNotPageableError: Operation cannot be paginated: list_rules`

**Comment by mjmayer** [comment]
@DaveHewy I'm seeing the same issue when creating a new WAF rule. My initial testing was with creating a WAF Condition. I will investigate further.

**Comment by mjmayer** [comment]
@DaveHewy I fixed one of the pagination errors in the aws_waf_rule. It won't work correctly at the moment if you have more than 100 rules. 

https://github.com/mjmayer/ansible/tree/aws_waf_region

**Comment by davehewy** [comment]
@mjmayer 
taken a look at the WIP branch.

Here's a simple paginator snippet that works well enough;

```
resp = client.list_rules() 
rules = []
while resp:
    rules += resp['Rules']
    resp = client.list_rules(NextMarker=resp['NextMarker']) if 'NextMarker' in resp else None
```

Also if i'd consider dropping;

`             cloudfront=dict(type='bool', default=True), `

in favour of 

`             waf_regional=dict(type='bool', default=True), `

or something thats service agnostic and reflective of the terminology used through the GUI for global vs regional.

Left a few inline comments as well. Let me know what you think.

**Comment by davehewy** [comment]
@mjmayer there is also large scale changes required under aws_waf_rule.py

around these lines:

```
        all_conditions[condition_type] = dict()
        try:
            paginator = client.get_paginator(method)
```

**Comment by mjmayer** [comment]
@DaveHewy Thanks for all the suggestions! The pagination code you provided will speed development up for me. I probably won't get to most of this until friday.

**Comment by davehewy** [comment]
@mjmayer I now have this all rudimentally working. If of any interest let me know.

**Comment by mjmayer** [comment]
@davehewy I have a working module as well, I'm just about to start writing the integration tests before I initiate a PR.

I'm interested in seeing your code to see where I could optimize mine.

**Comment by davehewy** [comment]
@mjmayer ill raise visibility of it shortly. 
im actually struggling with a nice way to keep the ansible/module_utils/aws/waf.py mods anyway. seems I can't import submodules relative to the custom ansible module i've created. nor can I find a reliable way of injecting files without horrible pre-tasks in ansible copying/installing deps on the host.

**Pull request closed without merging**